### PR TITLE
Workaround for noinit interfaces deprecated in SYCL2020

### PR DESCRIPTION
**Pull request by dparanic**

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h [hunk 1]
```diff
@@ -18,6 +18,8 @@
 
 #include <CL/sycl.hpp>
 
+#define __SYCL_2020_NOINIT_MACRO (__LIBSYCL_MAJOR_VERSION >= 5 && __LIBSYCL_MINOR_VERSION >= 2)
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h:21]: Could you please point where you found that actual values for macros are 5 and 2?

File: test/support/utils_sycl.h [hunk 2]
```diff
@@ -36,6 +36,20 @@
 namespace TestUtils
 {
 
+#if __cplusplus >= 201703L
+#   if __SYCL_2020_NOINIT_MACRO
+#       define __noinit sycl::no_init
+#   else
+#       define __noinit sycl::noinit
```
Comment (capatober) [inline test/support/utils_sycl.h:43]: Could you please explain why it was implemented using macros? Why it's not using, for example?

File: test/support/utils_sycl.h [hunk 3]
```diff
@@ -36,6 +36,20 @@
 namespace TestUtils
 {
 
+#if __cplusplus >= 201703L
+#   if __SYCL_2020_NOINIT_MACRO
+#       define __noinit sycl::no_init
```
Comment (capatober) [inline test/support/utils_sycl.h:41]: I wouldn't suggest to use `__`. This is mostly for internal using. However, it's not a blocking comment

File: test/general/buffer_wrapper.pass.cpp [hunk 4]
```diff
@@ -48,13 +48,13 @@ main()
...
+    auto data_ptr = sycl::host_accessor<uint32_t, 1, sycl::access_mode::read>(buf).get_pointer();
 
-    test(oneapi::dpl::begin(buf), oneapi::dpl::end(buf), begin, size);
-    test(oneapi::dpl::begin(buf, sycl::write_only), oneapi::dpl::end(buf, sycl::write_only), begin, size);
-    test(oneapi::dpl::begin(buf, sycl::write_only, sycl::noinit), oneapi::dpl::end(buf, sycl::write_only, sycl::noinit), begin, size);
-    test(oneapi::dpl::begin(buf, sycl::noinit), oneapi::dpl::end(buf, sycl::noinit), begin, size);
+    test(oneapi::dpl::begin(buf), oneapi::dpl::end(buf), data_ptr, size);
+    test(oneapi::dpl::begin(buf, sycl::write_only), oneapi::dpl::end(buf, sycl::write_only), data_ptr, size);
+    test(oneapi::dpl::begin(buf, sycl::write_only, __noinit), oneapi::dpl::end(buf, sycl::write_only, __noinit), data_ptr, size);
+    test(oneapi::dpl::begin(buf, __noinit), oneapi::dpl::end(buf, __noinit), data_ptr, size);
```
Comment (capatober) [inline test/general/buffer_wrapper.pass.cpp:56]: I'd suggest to test both: `sycl::no_init` and `sycl::property::noinit`. We need to verify that both options work

File: include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h [hunk 5]
```diff
@@ -156,14 +158,24 @@ __internal::sycl_iterator<Mode, T, Allocator> begin(sycl::buffer<T, /*dim=*/1, A
 
 template <typename T, typename Allocator, access_mode Mode>
 __internal::sycl_iterator<__internal::_ModeConverter<Mode>::__value, T, Allocator>
-    begin(sycl::buffer<T, /*dim=*/1, Allocator> buf, sycl::mode_tag_t<Mode>, sycl::property::noinit)
+    begin(sycl::buffer<T, /*dim=*/1, Allocator> buf, sycl::mode_tag_t<Mode>,
+#if __SYCL_2020_NOINIT_MACRO
+          sycl::property::no_init)
```
Comment (capatober) [inline include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h:163]: I think it needs to fix the Spec according to new changes as well (https://github.com/oneapi-src/oneAPI-spec/blob/main/source/elements/oneDPL/source/pstl.rst#buffer-wrappers).

Comment (capatober) [reply include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h:163]: Sure. Just to not forget about it.

File: test/support/utils_sycl.h (hunk 2 above)
Comment (capatober) [reply test/support/utils_sycl.h:43]: Ok. Why it's not a `constexpr` variable?
I mean, why not to have `constexpr auto noinit = sycl::noinit;` insead of `#define __noinit sycl::noinit`?

Comment (dparanic) [reply test/support/utils_sycl.h]: `sycl::no_init` is inline constexpr variable. So we can not use it with version of standard under 17. Also it's better to make constexpr constant inline but we can not do it with version of standard under 17. So I prefer to left it as a macro.

File: include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h (hunk 1 above)
Comment (dparanic) [reply include/oneapi/dpl/pstl/hetero/dpcpp/sycl_iterator.h]: Got information here: https://intel.github.io/llvm-docs/doxygen/version_8hpp.html

File: test/general/buffer_wrapper.pass.cpp (hunk 4 above)
Comment (dparanic) [reply test/general/buffer_wrapper.pass.cpp]: Added.

**Review by capatober — APPROVED** [review]
Looks good to me

**Pull request merged**